src/cache_structs_and_traits/cacher.rs

```diff
@@ -0,0 +1,246 @@
+            type AsBytes<'a> = Vec<u8>;
+
+            fn fixed_width() -> Option<usize> {
+                None
+            }
+
+            fn from_bytes<'a>(data: &[u8]) -> Self::SelfType<'a> {
+                serde_json::from_slice(data).expect("Could not convert to bytes.")
+            }
+
+            fn as_bytes<'a, 'b: 'a>(value: &'a Self::SelfType<'b>) -> Self::AsBytes<'a>
+            where
+                Self: 'b,
+            {
+                serde_json::to_vec(value).unwrap()
+            }
+
+            fn type_name() -> TypeName {
+                TypeName::new(type_name::<$type_name>())
+            }
+        }
+    };
+}
+
+implement_value_for_local_type!(GeneDoc);
+
+implement_value_for_local_type!(HgvsVariant);
+
+// for<'a> Self: From<Self::SelfType<'a>> is required so that cache_entry.value().into() works
+// for<'a> Self: Borrow<Self::SelfType<'a>> is required so that table.insert(key, object_to_cache.clone())?; works
+pub trait Cacheable: Sized + Clone + Value + 'static
+where
+    for<'a> Self: From<Self::SelfType<'a>>,
+    for<'a> Self: Borrow<Self::SelfType<'a>>,
+{
+    fn keys(&self) -> Vec<&str>;
+
+    fn table_definition() -> TableDefinition<'static, &'static str, Self> {
+        TableDefinition::new(type_name::<Self>())
+    }
+}
+
+impl Cacheable for HgvsVariant {
+    fn keys(&self) -> Vec<&str> {
+        vec![self.transcript_hgvs()]
+    }
+}
+
+impl Cacheable for GeneDoc {
+    fn keys(&self) -> Vec<&str> {
+        vec![self.symbol(), self.hgnc_id()]
+    }
+}
+
+pub struct Cacher<T: Cacheable> {
+    cache_file_path: PathBuf,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: Cacheable> Default for Cacher<T> {
+    fn default() -> Self {
+        let pkg_name = env!("CARGO_PKG_NAME");
+
+        let phenox_cache_dir = ProjectDirs::from("", "", pkg_name)
+            .map(|project_dir| project_dir.cache_dir().to_path_buf())
+            .or_else(|| home_dir().map(|home| home.join(pkg_name)))
+            .unwrap_or_else(|| panic!("Could not find cache directory or home directory."));
+
+        if !phenox_cache_dir.exists() {
+            fs::create_dir_all(&phenox_cache_dir)
+                .expect("Failed to create default cache directory.");
+        }
+
+        Cacher::new(phenox_cache_dir.join(type_name::<T>()))
+    }
+}
+
+impl<T: Cacheable> Cacher<T> {
+    pub fn new(cache_file_path: PathBuf) -> Self {
+        Cacher {
+            cache_file_path,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn cache_file_path(&self) -> &PathBuf {
+        &self.cache_file_path
+    }
+
+    pub fn init_cache(&self) -> Result<(), CacherError> {
+        let cache = RedbDatabase::create(self.cache_file_path.clone())?;
+
+        let write_txn = cache.begin_write()?;
+        {
+            write_txn.open_table(T::table_definition())?;
+        }
+        write_txn.commit()?;
+        Ok(())
+    }
+
+    pub fn open_cache(&self) -> Result<RedbDatabase, DatabaseError> {
```

Comment:
The Cacher implementation works, but is not thread safe.
This worked for our purposes, but other people might depend on it. If you do not want to make it thread safe, then you need to disclose it somewhere.

Reply:
Think about making this pub(crate).
Your public api should usually not return any foreign types.

```suggestion
    pub(crate) fn open_cache(&self) -> Result<RedbDatabase, DatabaseError> {
```

Or just remove it and call the content of the function where its needed.